### PR TITLE
Introduce interactive ArgoCD source plugin

### PR DESCRIPTION
**Pull request by pkosiec**
## Description

Changes proposed in this pull request:

- Introduce ArgoCD source plugin
   - Configure ArgoCD notifications on plugin startup
   - Receive interactive notifications from ArgoCD
   - Define triggers and templates based on ArgoCD notification catalog

## Questions

- Cleanup? For now it is manual. Should we introduce a new method for plugins and run it on SIGINT (context cancel)? I'm afraid it can be difficult to fit in time with modifying ConfigMaps and removing annotation for apps for all configurations 🤔 

## Follow-up PRs

- 1. Improvements 
    - [ ] Enhanced suport for other communication platforms
    - [x] Introduce JSON schemas for configuration and request (done in this PR)
- 2. Documentation

## Video

https://github.com/kubeshop/botkube/assets/7155799/cba259c5-39a8-4ba9-95f4-c1a3b69dc219

(sorry for a long delay when typing admin password - I was looking for it 😅)

BTW just two `kubectl` commands are supported:
image

## Testing

Checkout PR:
```
gh pr checkout 1220
```

Build all plugins:
```
rm -rf /tmp/botkube;  rm /tmp/botkube.yaml; rm -rf /tmp/plugins
PLUGIN_DOWNLOAD_URL_BASE_PATH="http://localhost:64538/plugin-dist" make gen-plugins-index
npx serve --listen 64538
```

Prepare `comm_config.yaml`:
```yaml
communications:
  "default-group":
    socketSlack:
      enabled: true
      appToken: "xapp-"
      botToken: "xoxb-"
      channels:
        "default":
          name: "botkube-demo"
          bindings:
            executors:
              - k8s-default-tools
            sources:
              - argocd

plugins:
  repositories:
    botkube:
       url: http://localhost:64538/plugins-index.yaml

  incomingWebhook:
    enabled: true
    inClusterBaseURL: http://host.k3d.internal:2115

sources:
  'argocd':
    botkube/argocd:
      enabled: true
      context:
        rbac:
          group:
            type: Static
            static:
              values: [*argocd-group]
      config:
        log:
          level: debug
        defaultSubscriptions:
          applications:
            - name: guestbook # your app name
              namespace: argocd

        argoCD:
          uiBaseUrl: http://localhost:8080 # if argo is not exported
```
For full options check out `values.yaml`.

Apply these RBAC resources:

<details><summary>Details</summary>
<p>

```yaml
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: argo-rbac
rules:
  - apiGroups: [""]
    resources: ["configmaps"]
    verbs: ["get", "update"]
  - apiGroups: ["argoproj.io"]
    resources: ["applications"]
    verbs: ["get", "patch"]
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: argo-rbac
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: argo-rbac
subjects:
- kind: Group
  name: argocd
  apiGroup: rbac.authorization.k8s.io
```

</p>
</details> 

Go through [ArgoCD Getting started](https://argo-cd.readthedocs.io/en/stable/getting_started/) doc and install Argo. Do all steps until step 7. Do not sync the app yet.

Run Botkube locally:

```go
export KUBECONFIG=...
export BOTKUBE_SETTINGS_LOG_LEVEL=debug           
export BOTKUBE_PLUGINS_CACHE__DIR="/tmp/plugins"
export BOTKUBE_CONFIG_PATHS="$(pwd)/helm/botkube/values.yaml,$(pwd)/comm_config.yaml"
export BOTKUBE_SETTINGS_KUBECONFIG="${KUBECONFIG}"
export LOG_LEVEL_SOURCE_BOTKUBE_ARGOCD="debug"
go run ./cmd/botkube-agent/main.go
```

After startup, wait for Argo notification. Sync (deploy the app) according to the guide: https://argo-cd.readthedocs.io/en/stable/getting_started/#7-sync-deploy-the-application

Delete the app

```
argocd app delete guestbook
```

## Related issue(s)

#1175

### CODE REVIEW

File: helm/botkube/values.yaml [hunk 1]
```diff
@@ -46,11 +46,25 @@ rbac:
...
+        - apiGroups: ["*"]
+          resources: ["*"]
+          verbs: ["get", "watch", "list"]
+    - name: &argocd-group "argocd"
+      rules:
+        - apiGroups: [""]
+          resources: ["configmaps"]
+          verbs: ["get", "update"]
+        - apiGroups: ["argoproj.io"]
+          resources: ["applications"]
```
Comment (huseyinbabal) [inline helm/botkube/values.yaml:66]: Better to check how this behaves if customer never installed ArgoCD CRDs, and just tries to use Botkube only. Will it throw "resource not found"?

File: internal/source/argocd/argocd.go [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+	"github.com/kubeshop/botkube/pkg/pluginx"
+)
+
+var _ source.Source = (*Source)(nil)
+
+//go:embed jsonschema.json
+var jsonSchema string
+
+const (
+	// PluginName is the name of the Kubernetes Botkube plugin.
```
Comment (huseyinbabal) [inline internal/source/argocd/argocd.go:29]:
```suggestion
	// PluginName is the name of the ArgoCD Botkube plugin.
```

File: internal/source/argocd/argocd.go [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+	}
+}
+
+type subscription struct {
+	TriggerName string
+	WebhookName string
+	Application config.K8sResourceRef
+}
+
+// Stream streams Kubernetes events
```
Comment (huseyinbabal) [inline internal/source/argocd/argocd.go:56]:
```suggestion
// Stream streams ArgoCD events
```

File: internal/source/argocd/argocd.go [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+var jsonSchema string
+
+const (
+	// PluginName is the name of the Kubernetes Botkube plugin.
+	PluginName = "argocd"
+
+	description = "Argo source plugin is used to watch ArgoCD events."
+)
+
+// Source Kubernetes source plugin data structure
```
Comment (huseyinbabal) [inline internal/source/argocd/argocd.go:35]:
```suggestion
// Source ArgoCD source plugin data structure
```

File: internal/source/argocd/argocd.go [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+	}
+
+	return source.ExternalRequestOutput{
+		Event: source.Event{
+			Message: msg,
+		},
+	}, nil
+}
+
+// Metadata returns metadata of Kubernetes configuration
```
Comment (huseyinbabal) [inline internal/source/argocd/argocd.go:133]:
```suggestion
// Metadata returns metadata of ArgoCD configuration
```

File: internal/source/argocd/jsonschema.json
Comment (huseyinbabal) [inline internal/source/argocd/jsonschema.json:1]: This file is empty, maybe forgot to push actual content? Or should it be in gitignore?

File: internal/source/argocd/render.go [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+)
+
+func renderStringIfTemplate(tpl string, srcCtx source.CommonSourceContext) (string, error) {
+	if !strings.Contains(tpl, goTplOpeningTag) {
+		return tpl, nil
+	}
+
+	tmpl, err := gotemplate.New("tpl").Parse(tpl)
+	if err != nil {
+		return "", err
```
Comment (huseyinbabal) [inline internal/source/argocd/render.go:23]:
```suggestion
		return "", fmt.Errorf("while parsing  string %q: %w", tpl, err)
```

File: pkg/config/testdata/TestLoadConfigSuccess/config.golden.yaml [hunk 7]
```diff
@@ -175,3 +175,4 @@ plugins:
     incomingWebhook:
         enabled: false
         port: 0
+        inClusterBaseURL: ""
```
Comment (huseyinbabal) [inline pkg/config/testdata/TestLoadConfigSuccess/config.golden.yaml:178]: Does that mean "In-cluster Base URL", or "Incoming cluster Base URL". I think this is used in slack buttons to go to ArgoCD dashbord, maybe better to state like "publicClusterBaseURL" or something.

File: internal/source/argocd/jsonschema.json
Comment (pkosiec) [reply internal/source/argocd/jsonschema.json:1]: Right, I didn't add the schema yet, will do in one of the follow-ups probably, so in the meantime I'll remove the file 👍

File: helm/botkube/values.yaml (hunk 1 above)
Comment (pkosiec) [reply helm/botkube/values.yaml:66]: Right, actually I'll introduce `enabled` property so that we don't create not used resources 👍

**Review by huseyinbabal — APPROVED** [review]
Works as expected 🚀 I have a couple of comments;
- The minor ones on PR
- The buttons beside `get` `describe` dropdown seems, they are tightly coupled with dropdown, but actually they are not. For example, I selected `get` and tried to click `Details` to run actual command, however it is automatically executed once selected. Maybe we can rename `Details` as `View in ArgoCD`, but not so important.
- Tested on discord too, works great
Screenshot 2023-09-06 at 14 20 46
Screenshot 2023-09-06 at 14 20 59

File: internal/source/argocd/jsonschema.json
Comment (huseyinbabal) [reply internal/source/argocd/jsonschema.json:1]: Actually, I liked the embed notation, working in schema json more easy than trying to edit json in single quotes without breaking the indent.

**Comment by pkosiec** [comment]
@huseyinbabal Thank you Huseyin! the button/dropdown layout is the same as in Kubernetes source, so yeah, they are not coupled. Maybe I'll make the primary button a default one? And yes, I'll change labels. Thank you!

File: helm/botkube/values.yaml [hunk 8]
```diff
@@ -53,11 +55,13 @@ rbac:
   # -- Use this to create additional RBAC resources for specified groups.
   groups:
     - name: &static-group-name "botkube-plugins-default"
+      create: true
       rules:
         - apiGroups: ["*"]
           resources: ["*"]
           verbs: ["get", "watch", "list"]
     - name: &argocd-group "argocd"
+      create: false # change to true when using ArgoCD plugin
```
Comment (huseyinbabal) [inline helm/botkube/values.yaml:64]: I think this part is important, but I think inline (the comment which beside the config line) does not reflects helm chart readme. Do you think we should make it like 
```yaml
....
## -- change to true when using ArgoCD plugin
create: false
...
```

Comment (pkosiec) [reply helm/botkube/values.yaml]: Agree, pushed fix: https://github.com/kubeshop/botkube/pull/1220/commits/b464ea8c71caaf644e99f9183d3de6edc90381c1

**Pull request merged**